apis/v1alpha1/destination.go

```diff
@@ -43,18 +43,18 @@ func (current *Destination) Validate(ctx context.Context) *apis.FieldError {
 
 func ValidateDestination(dest Destination) *apis.FieldError {
 	if dest.Ref == nil && dest.URI == nil {
-		return apis.ErrGeneric("expected at least one, got neither", "uri", "[apiVersion, kind, name]")
+		return apis.ErrGeneric("expected at least one, got neither", "ref", "uri")
 	}
 	if dest.Ref != nil && dest.URI != nil && dest.URI.URL().IsAbs() {
-		return apis.ErrGeneric("URI with absolute URL is not allowed when Ref exists", "uri", "[apiVersion, kind, name]")
-	}
-	if dest.Ref != nil && dest.URI == nil{
-		return validateDestinationRef(*dest.Ref)
+		return apis.ErrGeneric("Absolute URI is not allowed when Ref is present", "ref", "uri")
 	}
 	// IsAbs check whether the URL has a non-empty scheme. Besides the non non-empty scheme, we also require dest.URI has a non-empty host
 	if dest.Ref == nil && dest.URI != nil && (!dest.URI.URL().IsAbs() || dest.URI.Host == "") {
-			return apis.ErrInvalidValue("URI with Relative URL is not allowed when Ref is absent", "uri")
+			return apis.ErrInvalidValue("Relative URI is not allowed when Ref is absent",  "uri")
 		}
+	if dest.Ref != nil && dest.URI == nil{
```

Comment:
I think this if statement isn't necessary. There are 4 options for the state of these two fields:
1. ref nil, uri nil
1. ref not nil, uri not nil
1. ref nil, uri not nil
1. ref not nil, uri nil

We've already checked the first 3, so the only possibility at this point is 4, and there's no need for an if (or the `return nil` at the end)

Reply:
If we remove the 4th and remove return nil. The logic doesn't work.
For example, with destination with only valid URI, Ref is absent.
It pass the first three check. When there is no 4th check and return nil, it will try to call 
`validateDestinationRef(*dest.Ref)`. However, the Ref of destination is nil,  we will have error 'invalid pointer found on stack'. The prerequisites to call `validateDestinationRef(*dest.Ref)` is the Ref of destination should be present

Reply:
You're right, I missed that there were additional restrictions on cases 2 and 3. Ignore this comment!